pkg/virt-api/rest/rest.go

```diff
@@ -1,17 +1,98 @@
 package rest
 
 import (
+	"flag"
+	"fmt"
 	"github.com/emicklei/go-restful"
+	"github.com/go-kit/kit/endpoint"
+	"golang.org/x/net/context"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api"
+	kubev1 "k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/pkg/fields"
+	"k8s.io/client-go/pkg/labels"
+	"k8s.io/client-go/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
 	"kubevirt.io/kubevirt/pkg/api/v1"
 	"kubevirt.io/kubevirt/pkg/healthz"
+	"kubevirt.io/kubevirt/pkg/middleware"
+	"kubevirt.io/kubevirt/pkg/rest/endpoints"
+	"strings"
 )
 
 var WebService *restful.WebService
+var spiceProxy string
 
 func init() {
 	WebService = new(restful.WebService)
 	WebService.Path("/").Consumes(restful.MIME_JSON).Produces(restful.MIME_JSON)
 	WebService.ApiVersion(v1.GroupVersion.String()).Doc("help")
 	restful.Add(WebService)
 	WebService.Route(WebService.GET("/apis/" + v1.GroupVersion.String() + "/healthz").To(healthz.KubeConnectionHealthzFunc).Doc("Health endpoint"))
+	// TODO should be reloadable, use configmaps and update on every access? Watch a config file and reload?
+	flag.StringVar(&spiceProxy, "spice-proxy", "", "Spice proxy to use when spice access is requested")
+}
+
+func NewSpiceSubResourceEndpoint(cli *rest.RESTClient, coreCli *kubernetes.Clientset, gvr schema.GroupVersionResource) endpoint.Endpoint {
+	return func(ctx context.Context, payload interface{}) (interface{}, error) {
+		metadata := payload.(*endpoints.Metadata)
+		result := cli.Get().Namespace(metadata.Namespace).Resource(gvr.Resource).Name(metadata.Name).Do()
+		if result.Error() != nil {
+			return nil, middleware.NewInternalServerError(result.Error())
+		}
+		obj, err := result.Get()
+		if err != nil {
+			return nil, middleware.NewInternalServerError(result.Error())
+		}
+
+		vm := obj.(*v1.VM)
+
+		if vm.Status.Phase != v1.Running {
+			return nil, middleware.NewResourceNotFoundError("VM is not running")
+		}
+
+		// TODO allow specifying the spice device. For now select the first one.
+		for _, d := range vm.Spec.Domain.Devices.Graphics {
+			if strings.ToLower(d.Type) == "spice" {
```

Comment:
Well adding other graphics devices is valid, we just don't support any fowarding magic for them.